backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/InvoiceCandBL.java

```diff
@@ -1549,131 +1550,129 @@ private void setApprovalForInvoicing(final Collection<I_C_Invoice_Candidate> inv
-				final UomId ilUomId = UomId.ofRepoId(il.getC_UOM_ID());
-				final ProductId ilProductId = ProductId.ofRepoId(il.getM_Product_ID());
-
-				final String note;
-				if (isCreditMemo)
+				for (final I_C_Invoice_Candidate icToLink : toLinkAgainstIl)
 				{
-					final IDocTypeDAO docTypeDAO = Services.get(IDocTypeDAO.class);
+					//
+					// Make sure invoice candidate is fresh before we actually use it (task 06162, also see javadoc of updateInvalid method)
+					InterfaceWrapperHelper.refresh(icToLink);
 
-					// task 08927
-					note = "@C_DocType_ID@=" + docTypeDAO.getById(invoice.getC_DocType_ID()).getName() + ", @IsCreditedInvoiceReinvoicable@=" + creditMemoReinvoicable;
-					if (creditMemoReinvoicable)
+					// final BigDecimal qtyInvoiced = getQtyToInvoice(icToLink); // why on earth should we use the ic's QtyToInvoice, as opposed to the Qty that *was* actually invoiced to start with?
+					final StockQtyAndUOMQty qtysInvoiced;
+					final UomId ilUomId = UomId.ofRepoId(il.getC_UOM_ID());
+					final ProductId ilProductId = ProductId.ofRepoId(il.getM_Product_ID());
+
+					final String note;
+					if (isCreditMemo)
 					{
-						qtysInvoiced = StockQtyAndUOMQtys
-								.create(il.getQtyInvoiced(), ilProductId, il.getQtyEntered(), ilUomId)
-								.negate(); // this will allow the user to re-invoice, just as if the credit memo was a reversal
+						final IDocTypeDAO docTypeDAO = Services.get(IDocTypeDAO.class);
+
+						// task 08927
+						note = "@C_DocType_ID@=" + docTypeDAO.getById(invoice.getC_DocType_ID()).getName() + ", @IsCreditedInvoiceReinvoicable@=" + creditMemoReinvoicable;
+						if (creditMemoReinvoicable)
+						{
+							qtysInvoiced = StockQtyAndUOMQtys
+									.create(il.getQtyInvoiced(), ilProductId, il.getQtyEntered(), ilUomId)
+									.negate(); // this will allow the user to re-invoice, just as if the credit memo was a reversal
+						}
+						else
+						{
+							qtysInvoiced = StockQtyAndUOMQtys.createZero(ilProductId, ilUomId);
+						}
 					}
 					else
 					{
-						qtysInvoiced = StockQtyAndUOMQtys.createZero(ilProductId, ilUomId);
+						note = "";
+						qtysInvoiced = StockQtyAndUOMQtys
+								.create(il.getQtyInvoiced(), ilProductId, il.getQtyEntered(), ilUomId); // the standard case
 					}
+					createUpdateIla(icToLink, il, qtysInvoiced, note);
+					// note: if an ILA is created, the icToLink is automatically invalidated via C_Invoice_Line_Alloc model validator
 				}
-				else
-				{
-					note = "";
-					qtysInvoiced = StockQtyAndUOMQtys
-							.create(il.getQtyInvoiced(), ilProductId, il.getQtyEntered(), ilUomId); // the standard case
-				}
-				createUpdateIla(icToLink, il, qtysInvoiced, note);
-				// note: if an ILA is created, the icToLink is automatically invalidated via C_Invoice_Line_Alloc model validator
-			}
 
-			//
-			// Unset ApprovalForInvoicing flag in case we credit memoed + reinvoiceable our invoice
-			if (isCreditMemo && creditMemoReinvoicable)
```

Comment:
note i was a bit nervous about removing this, but i really can't see what it was there for in the first place